### ISSUE TITLE
Opening nested dialog when the parent dialog doesn't have focus closes parent dialog

### ISSUE THREAD

**Issue by Dremora** [ariakit/ariakit#3522]
### Current behavior

This may seem like a contrived example, but it has arisen from a real-world application. In order for the bug to happen, the following criteria need to be fulfilled:

- have two nested dialogs (nested in the React tree)
- outer dialog needs to have `autoFocusOnShow` set to `false`
- outer dialog needs to have an element (e.g. button) that opens inner dialog. This element should have `onMouseDown={(e) => e.preventDefault()}`. The reason for this specific code is that in our application this interactive element comes from [@hello-pangea/dnd](https://github.com/hello-pangea/dnd), and according to their docs, [a draggable handle that also serves as a clickable element will have this behaviour](https://github.com/hello-pangea/dnd/blob/9d64c5cfd79f93931a90c3a7d1d8d2b5fb53aec8/docs/guides/how-we-use-dom-events.md#initial-mousedown) (to prevent the element from becoming focused)

### Steps to reproduce the bug

1. Open https://stackblitz.com/edit/cniyvy?file=dialog-nested%2Findex.tsx&theme=dark
2. Click "View Cart"
3. Click on any of the "Remove" buttons
4. Watch the dialog close

### Expected behavior

The outer dialog stays open, and the nested dialog appears.

### Workaround

In our case, removing `autoFocusOnShow={false}` is probably the best solution. However, it's not a proper workaround as it changes the behaviour.

### Possible solutions

According to my understanding, this happens when the outer dialog doesn't have focus at the moment the inner dialog opens. The focus is outside the dialog (e.g. on the `<body>` element), and so the focus travels from outside the dialogs directly to the nested dialog. There's likely code in Ariakit that incorrectly identifies such focus event as moving focus outside of the outer dialog. Placing `hideOnInteractOutside` with an event handler on the outer dialog demonstrates that it's this focus event that triggers outer dialog to be closed.

**Comment by diegohaz** [comment]
Yes, the dialog element must be focused at least once to trigger the check on nested dialogs:

https://github.com/ariakit/ariakit/blob/6d29ca1279de5c5da44c89d0242b03f93918f055/packages/ariakit-react-core/src/dialog/utils/use-hide-on-interact-outside.ts#L90-L94

Could you provide more details on why you need `autoFocusOnShow={false}` on the outer dialog?

**Comment by Dremora** [comment]
> Could you provide more details on why you need autoFocusOnShow={false} on the outer dialog?

The only reason: depending on what opened the dialog, the first focusable element may have `:focus-visible` pseudoclass applied, which results in a focus ring. We only prefer to display focus ring on buttons when keyboard interaction was involved.

**Comment by diegohaz** [comment]
> _Update 2_: there's another reason (although not applicable to this particular dialog). We have an element with `autofocus` attribute within a dialog, but it doesn't work with `unmountOnHide`. Disabling `autoFocusOnShow` makes this element autofocus without any additional code.

You may want to use Ariakit's [`autoFocus`](https://ariakit.org/reference/focusable#autofocus) rather than the React `autofocus` prop. The former should work with any Ariakit dialog:

```diff
- <div autofocus />
+ <Focusable autoFocus render={<div />} />
```
